### PR TITLE
Restrict folding of stable functions during planning

### PR DESCRIPTION
**Pull request by asimrp**
Stable functions need not be folded in case the query is simple enough (e.g. pl/pgsql simple expression, or "select foo()").

Without this patch, every time a simple expression is evaluated, a new plan and a new ExprState are generated. Because we only clean up the expression states created for simple expressions at the end of transaction, this will result in millions of expression states, one per tuple, being generated that are not reused after they are initially created.

This patch is taken from Heikki's review comment: https://github.com/greenplum-db/gpdb/pull/5947#issuecomment-428103480

This patch also removes GPDB specific update for the test `guc.sql`. Prior to this commit, since we were aggressively folding stable functions, `current_setting()` was evaluated at plan time causing a diff from upstream. This patch realigns the test with upstream.

Thanks to @d for valuable guidance and carving out the test case.

We plan to close PR #5947 in favor of this one.

### CODE REVIEW

**Comment by joaopapereira** [comment]
At a first glance the code looks like it is doing what is supposed. 
Nevertheless I do think it would be more interesting to extract that if statement into a function like:
```
static bool need_to_eval_const_expressions(eval_const_expressions_context *context)
{
    return context->root && context->root->parse->rtable && context->root->glob;
}
```

**Comment by asimrp** [comment]
Thank you for the review @joaopapereira.  My concern with your suggestion is that this code path is taken per tuple in some cases.  Introducing a function call in this path risks performance.  I'm sure the simple function you've suggested will be inlined but why take changes.

This patch introduces a regression in case a stable function is used in a subquery.  E.g. consider this query from qp_functions_in_subquery.sql:
```SQL
SELECT * FROM foo, (SELECT func1_read_int_sql_stb(func2_read_int_stb(5))) r order by 1,2,3;
```
The function func2_read_int_stb is defined as:
```SQL
CREATE FUNCTION func2_read_int_stb(x int) RETURNS int AS $$
DECLARE
    r int;
BEGIN
    SELECT d FROM bar WHERE c = $1 LIMIT 1 INTO r;
    RETURN r;
END
$$ LANGUAGE plpgsql STABLE READS SQL DATA;
```
The patch no longer folds func2_read_int_stb() into a const because the subquery doesn't have a range table defined.  As a result, the query fails during execution:
```
ERROR query plan with multiple segworker groups is not supported
likely caused by a function that reads or modifies data in a distributed table
```

File: src/backend/optimizer/util/clauses.c [hunk 1]
```diff
@@ -4338,18 +4355,9 @@ evaluate_function(Oid funcid, Oid result_type, int32 result_typmod,
...
-		 * stable functions in any case. But it means that the plan is only
-		 * good for this execution, and will need to be re-planned on next one.
-		 * For GPDB, that's considered a good tradeoff, as typical queries are
-		 * long running, and evaluating the stable functions aggressively can
-		 * allow partition pruning to happen, which can be a big win.
-		 */
+		/* okay, but we cannot reuse this plan */
 		context->root->glob->oneoffPlan = true;
 	}
 	else
```
Comment (paul-guo-) [inline src/backend/optimizer/util/clauses.c]: On question: So with this patch, this function (`evaluate_function()`) returns NULL for "select foo()", and then won't PL/PGSQL replan for "select foo()" for each tuple?

**Comment by hlinnaka** [comment]
> This patch introduces a regression in case a stable function is used in a subquery. E.g. consider this query from qp_functions_in_subquery.sql:
> 
> ```sql
> SELECT * FROM foo, (SELECT func1_read_int_sql_stb(func2_read_int_stb(5))) r order by 1,2,3;
> ```
> 
> The function func2_read_int_stb is defined as:
> 
> ```sql
> CREATE FUNCTION func2_read_int_stb(x int) RETURNS int AS $$
> DECLARE
>     r int;
> BEGIN
>     SELECT d FROM bar WHERE c = $1 LIMIT 1 INTO r;
>     RETURN r;
> END
> $$ LANGUAGE plpgsql STABLE READS SQL DATA;
> ```
> 
> The patch no longer folds func2_read_int_stb() into a const because the subquery doesn't have a range table defined. As a result, the query fails during execution:
> 
> ```
> ERROR query plan with multiple segworker groups is not supported
> likely caused by a function that reads or modifies data in a distributed table
> ```

Hmm. What if we only do the range table check at the top level? Like this: (perhaps structure the checks differently, to make it more readable, though)

```
if (root && (root->parse->rtable || root->query_level > 1) && root->glob)
    context.eval_stable_functions = true;
else
    context.eval_stable_functions = false;
return eval_const_expressions_mutator(node, &context);
```
I think that would cover the above example. Are there other cases we need to worry about?

**Comment by asimrp** [comment]
That check will enable stable function folding for all subqueries.  But then we don't want that to happen for plpgsql simple expressions such as
```SQL
declare
   t int;
begin
   select * from (select stablefunc()) s into t;
...
```

**Comment by hlinnaka** [comment]
On 17/10/2018 20:02, Asim RP wrote:
> That check will enable stable function folding for all subqueries. But 
> then we don't want that to happen for plpgsql simple expressions such as
> 
> declare
>     tint;
> begin
>     select  *  from  (select  stablefunc()) s into t;
> ...

I believe that would not be treated as a "simple" expression in 
PL/pgSQL, because the outer query has a range table (containing one 
RTE_SUBQUERY entry).

- Heikki

**Comment by tvesely** [comment]
Actually, because the planner is smart enough to turn this query into a single result node, it is, indeed, considered to be a 'simple' expression by PL/pgSQL. 


```
pivotal=# explain verbose select * from (select now()) a;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Result  (cost=0.00..0.01 rows=1 width=0)
   Output: '2018-10-18 13:31:49.028701-07'::timestamp with time zone
 Planning time: 0.212 ms
 Optimizer: legacy query optimizer
 Settings: optimizer=off
(5 rows)

```

**Comment by hlinnaka** [comment]
On 18/10/2018 23:39, Taylor Vesely wrote:
> Actually, because the planner is smart enough to turn this query into
> a single result node, 

Doesn't matter. The checks in exec_simple_check_plan() run on the parsed 
Query, before planning.

> it is, indeed, considered to be a 'simple' expression by PL/pgSQL.
Try putting an elog(NOTICE) in exec_simple_recheck_plan, or use 
debugger, to verify.

- Heikki

**Comment by tvesely** [comment]
Thanks Heikki! I stand corrected. I was relying on the test in this PR to tell me if it was a simple expression, but the test doesn't actually check that- it only checks if we are evaluating stable functions more than once.

**Comment by khannaekta** [comment]
> Hmm. What if we only do the range table check at the top level? Like this: (perhaps structure the checks differently, to make it more readable, though)
> 
> ```
> if (root && (root->parse->rtable || root->query_level > 1) && root->glob)
>     context.eval_stable_functions = true;
> else
>     context.eval_stable_functions = false;
> return eval_const_expressions_mutator(node, &context);
> ```
> I think that would cover the above example. Are there other cases we need to worry about?

This _does_ cover the above example. It relies on the "query shape check" in pl/pgsql introduced in 9.2, which is merged into Greenplum 6. That means that this fix won't be applicable to Greenplum 5 (even if it cherry-picks cleanly).

**Comment by hlinnaka** [comment]
On 20/10/2018 01:33, khannaekta wrote:
>     Hmm. What if we only do the range table check at the top level? Like
>     this: (perhaps structure the checks differently, to make it more
>     readable, though)
> 
>     |if (root && (root->parse->rtable || root->query_level > 1) &&
>     root->glob) context.eval_stable_functions = true; else
>     context.eval_stable_functions = false; return
>     eval_const_expressions_mutator(node, &context); |
> 
>     I think that would cover the above example. Are there other cases we
>     need to worry about?
> 
> This /does/ cover the above example. It relies on the "query shape 
> check" in pl/pgsql introduced in 9.2, which is merged into Greenplum 6. 
> That means that this fix won't be applicable to Greenplum 5 (even if it 
> cherry-picks cleanly).

We could backport the query shape checks, too.

- Heikki

**Comment by hlinnaka** [comment]
>>  It relies on the "query shape check" in pl/pgsql introduced in 9.2, which is merged into Greenplum 6. That means that this fix won't be applicable to Greenplum 5 (even if it cherry-picks cleanly).

> We could backport the query shape checks, too.

Actually, thinking about this a bit more, I don't see why the query shape check in PL/pgSQL makes any difference. Does it matter, whether PL/pgSQL realizes that it's not going to be a "simple" plan, before or after planning?

I tried cherry-picking this to 5X_STABLE, and it worked fine. Passes the new test added to `bfv_planner`, with trivial adjustments to the error context message. Are we missing test coverage for some case where it's not working as we'd like, or is it in fact OK?

**Comment by hlinnaka** [comment]
On 24 October 2018 02:25:51 WEST, Jesse Zhang <notifications@github.com> wrote:
>> I tried cherry-picking this to 5X_STABLE
>
>Did it include the query level check you suggested in
>https://github.com/greenplum-db/gpdb/pull/5971#issuecomment-430506161 ?

No, I didn't include that.

>I don't believe that check will work

Why is that?


- Heikki

File: src/backend/optimizer/util/clauses.c [hunk 2]
```diff
@@ -2479,6 +2480,22 @@ eval_const_expressions(PlannerInfo *root, Node *node)
...
+	 * also aggressively evaluate stable functions, in addition to immutable
+	 * ones. Such plans cannot be reused, and therefore need to be re-planned
+	 * on every execution, but it can be a big win if it allows partition
+	 * elimination to happen. That's considered a good tradeoff in GPDB, as
+	 * typical queries are long-running.
+	 *
+	 * We also need 'root->glob' if we do that, so that we can mark the plan
+	 * as one-off, in root->glob->oneoffPlan.
+	 */
+	if (root && root->parse->rtable && root->glob)
```
Comment (melanieplageman) [inline src/backend/optimizer/util/clauses.c]: There is a problem with this. With Orca on, `eval_const_expressions` is not called (Orca does its own constant folding), so this `context.eval_stable_functions` is not set. However, with Orca on, `evaluate_function` is called and references this variable. This means this code doesn't work with Orca on.

**Comment by hlinnaka** [comment]
On 24/10/2018 18:22, Melanie wrote:
> There is a problem with this. With Orca on, `eval_const_expressions`
> is not called (Orca does its own constant folding), so this
> `context.eval_stable_functions` is not set. However, with Orca on,
> `evaluate_function` is called and references this variable. This
> means this code doesn't work with Orca on.

Hmm. I think you're saying that we need to add the same piece of code in 
fold_constants(), that this patch adds to eval_const_expressions(). Fair 
enough, let's do that.

- Heikki

File: src/backend/optimizer/util/clauses.c [hunk 3]
```diff
@@ -5729,3 +5739,32 @@ tlist_matches_coltypelist(List *tlist, List *coltypelist)
 
 	return true;
 }
+
+static bool
+planner_should_eval_stable_functions(PlannerInfo *root)
+{
```
Comment (hlinnaka) [inline src/backend/optimizer/util/clauses.c]: Why are the rules different for ORCA and the Postgres planner?

File: src/test/modules/test_planner/integration_tests/planner_integration_tests.c [hunk 4]
```diff
@@ -0,0 +1,97 @@
+#include "postgres.h"
+#include "tcop/tcopprot.h"
+
+#include "planner_integration_tests.h"
+
+#include "src/assertions.h"
+#include "src/planner_test_helpers.h"
+
+static void
+test_window_function_with_subquery_has_correct_extparams()
```
Comment (hlinnaka) [inline src/test/modules/test_planner/integration_tests/planner_integration_tests.c]: This test seems unrelated to the patch.

File: src/test/modules/test_planner/integration_tests/planner_integration_tests.c [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		is_equal_to(0));
+
+	assert_that_int(plannedStatement->nInitPlans,
+		is_equal_to(0));
+}
+
+static void
+test_vanilla_subquery_has_correct_extparams()
+{
+	const char *subquery_string = "select ( \
```
Comment (hlinnaka) [inline src/test/modules/test_planner/integration_tests/planner_integration_tests.c]: This also seems unrelated.

File: src/test/modules/test_planner/integration_tests/planner_integration_tests.c [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		is_equal_to(0));
+
+	assert_that_int(plannedStatement->nInitPlans,
+		is_equal_to(0));
+}
+
+static void
+test_stable_function_in_subquery_is_evaluated_to_const()
+{
+	const char *query_string = "select * from (select now()) a;";
```
Comment (hlinnaka) [inline src/test/modules/test_planner/integration_tests/planner_integration_tests.c]: I think these would be more straightforward to test as normal pg_regress tests. Do `EXPLAIN VERBOSE` on that, and verify that the function call is replaced with the constant.

That's actually problematic for now(), because that's value changes, but you can replace it with some other stable function.

Comment (hlinnaka) [reply src/test/modules/test_planner/integration_tests/planner_integration_tests.c]: Looking further down in the path, there actually seem to be pg_regress tests like that included. Not using EXPLAIN, but other mechanisms. So it seems we can just remove this test_planner module.

File: src/test/modules/test_planner/integration_tests/planner_integration_tests.c (hunk 4 above)
Comment (berlin-ab) [reply src/test/modules/test_planner/integration_tests/planner_integration_tests.c]: Good point. We'll remove them.

File: src/backend/optimizer/util/clauses.c (hunk 3 above)
Comment (tvesely) [reply src/backend/optimizer/util/clauses.c]: Excellent point- I just stupidly assumed that we couldn't make the same assumptions about the state of the `PlannerInfo` here that we made for planner. Fixing it now.

File: src/test/modules/test_planner/integration_tests/planner_integration_tests.c (hunk 6 above)
Comment (asimrp) [reply src/test/modules/test_planner/integration_tests/planner_integration_tests.c]: I agree, these tests are redundant and not adding much value.  They are equivalent to the other two pg_regress tests.  This PR is merged already but let this serve as a note to ourselves - let's not over-engineer tests.

File: src/backend/optimizer/util/clauses.c (hunk 1 above)
Comment (asimrp) [reply src/backend/optimizer/util/clauses.c]: @paul-guo- apologies for missing your question all this while.  Let's consider the PL/PGSQL statement `x := foo();`.  This is translated to `select foo();`.  If `foo()` is a stable function, with this patch, the expression will be planned as a `FuncExpr` and the plan will be cached.  For each tuple scanned within the same command, the `FuncExpr` will be evaluated but the cached plan will be reused.
